### ISSUE TITLE
Does not work if the given base port is already taken

### ISSUE THREAD

**Issue by kevgo** [indexzero/node-portfinder#24]
The test script below allocates port 3000, then calls portfinder with basePort set to 3000. 
With the old version 0.4.0 it works correctly, i.e. prints "1 2 3". 
With portfinder 1.0.0., it does not output "3", i.e. it never calls the callback given to `getPort`. 

```javascript
http = require('http');
portfinder = require('portfinder');
handler = function(_, res) {
  return res.end('existing server');
};

console.log(1);
server = http.createServer(handler).listen(3000, 'localhost', function() {
  console.log(2);
  portfinder.basePort = 3000;
  portfinder.getPort(function(err, port) {
    console.log(3);
  });
});
```

My setup:
* OS X 10.11.3
* Node v5.3.0

**Comment by winfinit** [comment]
i've encountered similar issue, with this pull request https://github.com/indexzero/node-portfinder/pull/25 above is printing "3"

**Comment by indexzero** [comment]
@kevgo does this work on `portfinder@0.4.0`?

**Comment by kevgo** [comment]
Yes, it works in 0.4.0. This issue is new in 1.0.0.

**Comment by winfinit** [comment]
@indexzero it would work, because of introduction of ::1, 0.0.0.0 and 127.0.0.1 test, basically previous bug was that if you start something on 0.0.0.0, and then test against 127.0.0.1, then same port would be reported as open, and that is actually correct in OSX, because 0.0.0.0 would listen as *.{port}, and * does not mean that listen on "EVERY" available interface, it is a default route for specified port, unless there was a bind on a specific device, so for example, this would successfully bind to all three interfaces:

```javascript
const http = require('http');

const port = process.argv[2] || 3000;

http.createServer( function(request, response) {
  response.writeHead(200, {'Content-Type': 'text/plain'});
  response.end('Hello World 0.0.0.0\n');
}).listen(port);

http.createServer( function(request, response) {
  response.writeHead(200, {'Content-Type': 'text/plain'});
  response.end('Hello World 127.0.0.1\n');
}).listen(port, "127.0.0.1");

http.createServer( function(request, response) {
  response.writeHead(200, {'Content-Type': 'text/plain'});
  response.end('Hello World ::1\n');
}).listen(port, "::1");
```

```
$ netstat -an |grep 3000
tcp6       0      0  ::1.3000               *.*                    LISTEN     
tcp4       0      0  127.0.0.1.3000         *.*                    LISTEN     
tcp46      0      0  *.3000                 *.*                    LISTEN   
```

and behavior would be as one would expect:
```
$ curl http://[::1]:3000 -g 6
Hello World ::1
$ curl http://127.0.0.1:3000
Hello World 127.0.0.1
$ curl http://0.0.0.0:3000
Hello World 127.0.0.1
$ curl http://192.168.191.221:3000
Hello World 0.0.0.0
```

however, this behavior creates confusion in most, because service creators assume that 0.0.0.0 is same as 127.0.0.1, or some assume that 0.0.0.0 will create a bind to all of the available interfaces

that is not true for linux operating system however, same server snippet would result in this:

```
listening on 3000
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: listen EADDRINUSE
    at exports._errnoException (util.js:746:11)
    at Server._listen2 (net.js:1156:14)
    at listen (net.js:1182:10)
    at net.js:1280:9
    at dns.js:85:18
    at process._tickCallback (node.js:355:11)
    at Function.Module.runMain (module.js:503:11)
    at startup (node.js:129:16)
    at node.js:814:3
```

**Comment by winfinit** [comment]
actually, similar scenarios could be on linux, however it seems that nodes "listen" behaves differently on linux vs OS X, for example, if you don't specify a host, node would create socket with protocol tcp6 and address would be :::3000, where if you specify host in listen like this: 'listen(3000, "0.0.0.0")' that would register host like this: "0.0.0.0:3000" with protocol set to tcp, and that would permit one to bind to ::1, so this can happen:

```
tcp        0      0 0.0.0.0:3000            0.0.0.0:*               LISTEN     
tcp6       0      0 ::1:3000                :::*                    LISTEN     
```

**Comment by eriktrom** [comment]
@kevgo - https://github.com/indexzero/node-portfinder/blob/master/lib/portfinder.js#L108

options.hosts should be option.host otherwise we recursively pass undefined from the 2nd recursion onward. this did reveal some other thoughts i have regarding making sure a host passed in does not resolve to an ip address that we by default look for, which is what would happen even if options.hosts was in fact option.host in that line mentioned, meaning while there was a bug, your use case would have still revealed another bug, namely, localhost resolves to `::1` on mac 10.10+ and (127.0.0.1 on 10.9 or less, same idea for any other OS - we need to check this sort of thing to get this right)

I'm repeating myself here a bit but wanted to cross post and summarize in case you or anyone has any spare cycles to help out with #25 if it does not handle all the use cases we need yet, thanks

**Comment by eriktrom** [comment]
fixed on master
